src/Confluent.SchemaRegistry.Serdes.Json/JsonSerializer.cs

```diff
@@ -0,0 +1,180 @@
+// Copyright 2020 Confluent Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Refer to LICENSE for more information.
+
+#pragma warning disable CS0618
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using NJsonSchema;
+using NJsonSchema.Validation;
+using Confluent.Kafka;
+
+
+namespace Confluent.SchemaRegistry.Serdes
+{
+    /// <summary>
+    ///     Json Serializer.
+    /// </summary>
+    /// <remarks>
+    ///     Serialization format:
+    ///       byte 0:           A magic byte that identifies this as a message with
+    ///                         Confluent Platform framing.
+    ///       bytes 1-4:        Unique global id of the JSON schema associated with
+    ///                         the data (as registered in Confluent Schema Registry),
+    ///                         big endian.
+    ///       following bytes:  The JSON data (utf8)
+    ///
+    ///     Internally, the serializer uses Newtonsoft.Json for
+    ///     serialization and NJsonSchema for schema creation and
+    ///     validation. You can use any property annotations recognised
+    ///     by these libraries.
+    ///
+    ///     Note: Off-the-shelf libraries do not yet exist to enable
+    ///     integration of System.Text.Json and JSON Schema, so this
+    ///     is not yet supported by the serializer.
+    /// </remarks>
+    public class JsonSerializer<T> : IAsyncSerializer<T>  where T : new()
+    {
+        private const int DefaultInitialBufferSize = 1024;
+
+        private bool autoRegisterSchema = true;
+        private int initialBufferSize = DefaultInitialBufferSize;
+        private SubjectNameStrategyDelegate subjectNameStrategy = null;
+        private ISchemaRegistryClient schemaRegistryClient;
+        
+        private HashSet<string> subjectsRegistered = new HashSet<string>();
+        private SemaphoreSlim serializeMutex = new SemaphoreSlim(1);
```

Comment:
Because there isn't a mutex construct that works with `async/await`. Well there wasn't when I did the avro implementation, and this is just copied over.